<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>cinet &mdash; cinet  documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=19f00094" />
      <link rel="stylesheet" type="text/css" href="_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
      <link rel="stylesheet" type="text/css" href="_static/graphviz.css?v=fd3f3429" />
      <link rel="stylesheet" type="text/css" href="_static/sphinx_paramlinks.css" />

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script src="_static/jquery.js?v=5d32c60e"></script>
        <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
        <script src="_static/documentation_options.js?v=5929fcd5"></script>
        <script src="_static/doctools.js?v=9a2dae69"></script>
        <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Example usage" href="example.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="#" class="icon icon-home">
            cinet
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="example.html">Example usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributing</a></li>
<li class="toctree-l1"><a class="reference internal" href="conduct.html">Code of Conduct</a></li>
<li class="toctree-l1"><a class="reference internal" href="autoapi/index.html">API Reference</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">cinet</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="#" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">cinet</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/index.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="cinet">
<h1>cinet<a class="headerlink" href="#cinet" title="Link to this heading"></a></h1>
<p>Scikit-Learn interface for CINET PyTorch siamese neural network.</p>
<p>DeepCINET is a deep “siamese” neural network architecture, where a contrastive loss function is used to learn feature weights that maximally discriminate relative response/target between valid pairs of training data. A hyper-parameter, delta, is used to define what a valid pair is by setting a minimum difference in response/target value for pairs to be included in model training, with the intuition that useful weights cannot be learned from samples that are too close together in response-space.</p>
<p>Concordance index is then used to assess rank accuracy. Concordance index was chosen because it is a non-parametric statistic that does not make
assumptions on data distributon or homoscedasticity. It can detect non-linear, monotonic associations.</p>
<p>ECINET is a one-dimensional neural network, which makes it essentially a linear regression model with regularization. It is comparable to model architectures like ElasticNet. It can be used
to assess if improved performance is delivered by the added complexity of DeepCINET.</p>
<p>Note, however, that siamese networks go hand-in-hand with few shot learning approaches. The idea is that features learned from large data in CINET can then be applied to learning done
on smaller real-world data in a transfer learning approach.</p>
<p>An initial implementation, trained on gene set expression data from cancer cell lines and meant to predict drug sensitivity rank, is available on the BHKLab’s public GitHub at https://github.com/bhklab/cinet.</p>
<section id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Link to this heading"></a></h2>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>$<span class="w"> </span>pip3<span class="w"> </span>install<span class="w"> </span>cinet
</pre></div>
</div>
</section>
<section id="usage">
<h2>Usage<a class="headerlink" href="#usage" title="Link to this heading"></a></h2>
<p>CINET can be used like any other Scikit-Learn model.</p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import CINET</span>
<span class="kn">from</span> <span class="nn">cinet</span> <span class="kn">import</span> <span class="o">*</span>

<span class="c1"># Create a DeepCINET model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">deepCINET</span><span class="p">()</span>
<span class="c1"># Or, create an ECINET model</span>
<span class="n">model</span> <span class="o">=</span> <span class="n">ECINET</span><span class="p">()</span>

<span class="c1"># Standard Scikit-Learn syntax</span>
<span class="n">model</span><span class="o">.</span><span class="n">fit</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">y</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">predict</span><span class="p">(</span><span class="n">X</span><span class="p">)</span>
<span class="n">model</span><span class="o">.</span><span class="n">score</span><span class="p">(</span><span class="n">X</span><span class="p">,</span><span class="n">y</span><span class="p">)</span>

<span class="c1"># You can use it with things like GridSearchCV easily</span>
<span class="n">GridSearchCV</span><span class="p">(</span><span class="n">deepCINET</span><span class="p">(</span><span class="n">device</span><span class="o">=</span><span class="s1">&#39;cpu&#39;</span><span class="p">,</span> <span class="n">batch_size</span><span class="o">=</span><span class="mi">2</span><span class="o">**</span><span class="mi">12</span><span class="p">),</span> <span class="n">param_grid</span><span class="p">,</span> <span class="n">refit</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">verbose</span> <span class="o">=</span> <span class="mi">3</span><span class="p">,</span><span class="n">n_jobs</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</section>
<section id="data-sources">
<h2>Data sources<a class="headerlink" href="#data-sources" title="Link to this heading"></a></h2>
<p>DeepCINET’s training datasets are composed of the Cancer Cell Line Encyclopedia (CCLE, https://www.orcestra.ca/pset/10.5281/zenodo.3905461), and the Cancer Therapeutics Response Portal (CTRP-v2, https://www.orcestra.ca/pset/10.5281/zenodo.7826870). On the other hand, the testing datasets include the Genentech Cell Line Screening Initiative (gCSI, https://www.orcestra.ca/pset/10.5281/zenodo.7829857) and the second version of the Genomics of Drug Sensitivity in Cancer (GDSC-v2, https://www.orcestra.ca/pset/10.5281/zenodo.5787145). All PSet R objects were downloaded from Orcestra (https://www.orcestra.ca/). An extra resource used during the execution of the project was the COSMIC Cancer Gene Census (https://cancer.sanger.ac.uk/census), to select genes related to cancer development. The CCLE, gCSI and GDSC-v2 datasets contain both RNA-Seq data as well as drug response (AAC) data, while the CTRP-v2 dataset exclusively contains drug response data.</p>
</section>
<section id="data-curation-process">
<h2>Data curation process<a class="headerlink" href="#data-curation-process" title="Link to this heading"></a></h2>
<p>The end goal is to generate <strong>one file for each drug</strong>, containing cell-lines’ AAC and RNA-Seq values. The curation process executed followed the following steps:</p>
<ol class="arabic simple">
<li><p>Extraction of <em>csv</em> files from PSets using R files. For CCLE, gCSI, and GDSC-v2, the <strong>OrcestraDatasetCuration.R</strong> code extracts RNA-Seq and AAC values in two separate <em>csv</em> files. For CTRP-v2, the code <strong>CTRP-curation.R</strong> extracts the experiments, drugs, and cell-lines <em>csv</em> files.</p></li>
<li><p>Obtention of the <strong>common drugs</strong> between the <strong>gCSI</strong> and <strong>GDSC-v2</strong> datasets.</p></li>
<li><p>Extraction from the CTRP-v2 experiments <em>csv</em> file of the list of cell-lines treated with any of the drugs obtained from step 2, as well as their AAC drug response data to that drug.</p></li>
<li><p>Intersection of <strong>common cell-lines</strong> between the <strong>CCLE</strong> and <strong>CTRP-v2</strong> datasets. For those cell-lines, concatenate RNA-Seq data from CCLE to CTRP-v2’s drug response data.</p></li>
<li><p>Obtention of <strong>common genes</strong> across the three final datasets, <strong>CCLE-CTRP-v2</strong>, <strong>gCSI</strong>, and <strong>GDSC-v2</strong>. Be careful with gene naming! Make sure to eliminate the <em>version</em> of the gene, i.e: (~~ENSG00000088038.18~~, ENSG00000088038)</p></li>
<li><p>Intersection of the genes obtained from step 5 with the ones in the COSMIC Cancer Gene Census.</p></li>
<li><p>Lastly, generation of <strong>one <em>csv</em> file per drug for each dataset</strong>, each containing cell-lines tested with that drug in that study. Each file contains AAC response values of those cell-lines to the drug at hand, as well as RNA-Seq gene expression data for Cancer Gene Census genes.</p></li>
</ol>
</section>
<section id="contributing">
<h2>Contributing<a class="headerlink" href="#contributing" title="Link to this heading"></a></h2>
<p>Interested in contributing? Check out the contributing guidelines. Please note that this project is released with a Code of Conduct. By contributing to this project, you agree to abide by its terms.</p>
</section>
<section id="license">
<h2>License<a class="headerlink" href="#license" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cinet</span></code> was created by Kevin Tabatabaei and Christopher Eeles. It is licensed under the terms of the MIT license.</p>
</section>
<section id="credits">
<h2>Credits<a class="headerlink" href="#credits" title="Link to this heading"></a></h2>
<p><code class="docutils literal notranslate"><span class="pre">cinet</span></code> was created with <a class="reference external" href="https://cookiecutter.readthedocs.io/en/latest/"><code class="docutils literal notranslate"><span class="pre">cookiecutter</span></code></a> and the <code class="docutils literal notranslate"><span class="pre">py-pkgs-cookiecutter</span></code> <a class="reference external" href="https://github.com/py-pkgs/py-pkgs-cookiecutter">template</a>.</p>
<div class="toctree-wrapper compound">
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="example.html" class="btn btn-neutral float-right" title="Example usage" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Kevin Tabatabaei, Christopher Eeles, Benjamin Haibe-Kains Lab (BHK Lab).</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>